src/core/test/fuzzing/src/test/scala/Fuzzing.scala

```diff
@@ -26,10 +29,53 @@ case class TestObject[S <: PipelineStage](stage: S,
 
 }
 
-trait FuzzingMethods {
-  def compareDFs(df1: DataFrame, df2: DataFrame): Boolean = {
-    df1.collect().toSet == df2.collect().toSet
+trait FuzzingMethods extends TestBase {
+  val epsilon = 1e-4
+  implicit lazy val doubleEq: Equality[Double] = TolerantNumerics.tolerantDoubleEquality(epsilon)
+  implicit lazy val dvEq: Equality[DenseVector] = new Equality[DenseVector]{
+    def areEqual(a: DenseVector, b: Any): Boolean = b match {
+      case bArr:DenseVector =>
+        a.values.zip(bArr.values).forall {case (x, y) => doubleEq.areEqual(x, y)}
+    }
+  }
+
+  implicit lazy val rowEq: Equality[Row] = new Equality[Row]{
+    def areEqual(a: Row, bAny: Any): Boolean = bAny match {
+      case b:Row =>
+        if(a.length != b.length){return false}
+        (0 until a.length).forall(j =>{
+          a(j) match {
+            case lhs: DenseVector =>
+              lhs === b(j)
+            case lhs: Double if lhs.isNaN =>
+              b(j).asInstanceOf[Double].isNaN
+            case lhs: Double =>
+              b(j).asInstanceOf[Double] === lhs
+            case lhs =>
+              lhs === b(j)
+          }
+        })
+    }
+  }
+
+  implicit lazy val dfEq: Equality[DataFrame] = new Equality[DataFrame]{
+    def areEqual(a: DataFrame, bAny: Any): Boolean = bAny match {
+      case ds:Dataset[_] =>
+        val b = ds.toDF()
+        if(a.columns !== b.columns){
+          return false
+        }
+        val aSort = a.collect()
+        val bSort = b.collect()
+        if (aSort.length != bSort.length){
+          return false
+        }
+        aSort.zip(bSort).forall {case (rowA, rowB) =>
```

Comment:
I don't think you can always do this -- the rows don't have any order in a dataframe.  So unless you do some expensive sort you can't really compare them like this.

Reply:
Fixed with a sort